libvirt/resource_libvirt_domain.go

```diff
@@ -494,6 +512,31 @@ func resourceLibvirtDomainCreate(d *schema.ResourceData, meta interface{}) error
 	return nil
 }
 
+// enableDHCP enables DHCP services on a given libvirt network
+func enableDHCP(networkDef defNetwork, network *libvirt.VirNetwork) error {
+	// mutex required due to concurrent reads of the network definition
+	// (adding the same DHCP range twice causes an error while creating the libvirt domain)
+	dhcpMutex.Lock()
+	defer dhcpMutex.Unlock()
```

Comment:
We need to lock around this entire function because multiple threads could both read that `ip.Dhcp == nil` and attempt to add a DHCP range, resulting in error. Also, I believe we need to read the libvirt network again inside the locks like I mentioned in the comment above.